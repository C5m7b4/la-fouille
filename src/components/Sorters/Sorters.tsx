import React from 'react';
import { ISorterProperty } from '../../interfaces/ISorterProperty';
import { PropsWithChildrenFunction } from '../../interfaces/PropsWithChildrenFunction';

import { ISortersProps } from '../../interfaces/ISorterProps';

function Sorters<T>(props: PropsWithChildrenFunction<ISortersProps<T>, T>) {
  //destructure our props
  const {
    dataSource,
    mainDivClassName,
    mainDivStyle,
    labelClassName,
    labelStyle,
    labelText = 'Sort by',
    selectClassName,
    selectStyle,
    selectOptionClassName,
    selectOptionStyle,
    setSortProperty,
  } = props;

  const object = dataSource.length > 0 ? dataSource[0] : {};

  return (
    <div style={mainDivStyle} className={mainDivClassName}>
      <label className={labelClassName} style={labelStyle}>
        {labelText}: &nbsp;
      </label>
      <select
        style={selectStyle}
        data-testid={'sorter-select'}
        className={selectClassName}
        onChange={(e) => {
          const values = e.target.value.split('-');
          if (values.length === 2) {
            setSortProperty({
              property: values[0] as any,
              isDescending: values[1] === 'true',
            });
          }
        }}
      >
        {Object.keys(object).map((key, i) => (
          <React.Fragment key={`rfs-${i}`}>
            <option
              style={selectOptionStyle}
              className={selectOptionClassName}
              data-testid={`ss-${key}-true`}
              key={`sorter-${i}-true`}
              value={`${key}-true`}
            >
              {key} descending
            </option>
            <option
              style={selectOptionStyle}
              className={selectOptionClassName}
              data-testid={`ss-${key}-false`}
              key={`sorter-${i}-false`}
              value={`${key}-false`}
            >
              {key} ascending
            </option>
          </React.Fragment>
        ))}
      </select>
    </div>
  );
}

export default Sorters;
