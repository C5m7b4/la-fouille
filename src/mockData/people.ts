import { IPerson } from '../interfaces/IPerson';

export const people: Array<IPerson> = [
  {
    firstName: 'Sawyer',
    lastName: 'Riggs',
    eyeColor: 'blue',
    birthday: new Date(
      'Sat Nov 12 1988 22:39:07 GMT-0600 (Central Standard Time)'
    ),
    isActive: false,
  },
  {
    firstName: 'Cindy',
    lastName: 'Robinson',
    eyeColor: 'hazel',
    birthday: new Date(
      'Wed Jul 03 1985 07:41:24 GMT-0500 (Central Daylight Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Candy',
    lastName: 'Dillon',
    eyeColor: 'green',
    birthday: new Date(
      'Tue Jan 10 1984 11:10:35 GMT-0600 (Central Standard Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Lisa',
    lastName: 'Finch',
    eyeColor: 'green',
    birthday: new Date(
      'Thu Feb 11 1982 00:45:06 GMT-0600 (Central Standard Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Latonya',
    lastName: 'Valentine',
    eyeColor: 'blue',
    birthday: new Date(
      'Sat Mar 08 1986 08:29:06 GMT-0600 (Central Standard Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Emily',
    lastName: 'Velez',
    eyeColor: 'hazel',
    birthday: new Date(
      'Sun Nov 29 1987 14:49:15 GMT-0600 (Central Standard Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Mullen',
    lastName: 'Mckee',
    eyeColor: 'green',
    birthday: new Date(
      'Fri Jun 09 1989 00:11:56 GMT-0500 (Central Daylight Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Molina',
    lastName: 'Barlow',
    eyeColor: 'hazel',
    birthday: new Date(
      'Thu May 01 1980 20:07:21 GMT-0500 (Central Daylight Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Phoebe',
    lastName: 'Castaneda',
    eyeColor: 'brown',
    birthday: new Date(
      'Sat Jan 18 1986 23:33:05 GMT-0600 (Central Standard Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Maldonado',
    lastName: 'Snow',
    eyeColor: 'brown',
    birthday: new Date(
      'Mon Jun 24 1985 15:21:42 GMT-0500 (Central Daylight Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Lucia',
    lastName: 'Gillespie',
    eyeColor: 'blue',
    birthday: new Date(
      'Wed Jan 20 1982 14:21:39 GMT-0600 (Central Standard Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Carrillo',
    lastName: 'Clay',
    eyeColor: 'brown',
    birthday: new Date(
      'Wed Nov 19 1980 18:50:34 GMT-0600 (Central Standard Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Nicholson',
    lastName: 'Chaney',
    eyeColor: 'brown',
    birthday: new Date(
      'Fri Jan 25 1980 02:49:55 GMT-0600 (Central Standard Time)'
    ),
    isActive: false,
  },
  {
    firstName: 'Leonard',
    lastName: 'Richmond',
    eyeColor: 'hazel',
    birthday: new Date(
      'Wed Oct 04 1989 17:10:05 GMT-0500 (Central Daylight Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Beatriz',
    lastName: 'Morrow',
    eyeColor: 'blue',
    birthday: new Date(
      'Fri Dec 13 1985 07:37:35 GMT-0600 (Central Standard Time)'
    ),
    isActive: false,
  },
  {
    firstName: 'Ina',
    lastName: 'Gilliam',
    eyeColor: 'hazel',
    birthday: new Date(
      'Sun Jun 27 1982 00:54:06 GMT-0500 (Central Daylight Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Janet',
    lastName: 'Mcclain',
    eyeColor: 'brown',
    birthday: new Date(
      'Sun Aug 21 1983 18:41:58 GMT-0500 (Central Daylight Time)'
    ),
    isActive: false,
  },
  {
    firstName: 'Parker',
    lastName: 'Jenkins',
    eyeColor: 'hazel',
    birthday: new Date(
      'Thu Nov 03 1988 04:00:53 GMT-0600 (Central Standard Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Bertie',
    lastName: 'Cruz',
    eyeColor: 'hazel',
    birthday: new Date(
      'Sat Sep 08 1984 09:01:43 GMT-0500 (Central Daylight Time)'
    ),
    isActive: true,
  },
  {
    firstName: 'Cooke',
    lastName: 'Kerr',
    eyeColor: 'hazel',
    birthday: new Date(
      'Wed May 11 1983 00:24:23 GMT-0500 (Central Daylight Time)'
    ),
    isActive: false,
  },
];
