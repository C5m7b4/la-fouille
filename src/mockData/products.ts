import { IProduct } from '../interfaces/IProduct';

import { ProductCategory } from '../enums/ProductCategory';

export const products: Array<IProduct> = [
  {
    id: 1,
    productName: 'T-Shirts - green',
    price: 499,
    size: 'medium',
    category: ProductCategory.CLOTHES,
    isRed: false,
    isBlue: false,
    isGreen: true,
  },
  {
    id: 2,
    productName: 'T-Shirts - red',
    price: 499,
    size: 'medium',
    category: ProductCategory.CLOTHES,
    isRed: true,
    isBlue: false,
    isGreen: false,
  },
  {
    id: 3,
    productName: 'T-Shirts - blue',
    price: 499,
    size: 'medium',
    category: ProductCategory.CLOTHES,
    isRed: false,
    isBlue: true,
    isGreen: false,
  },
  {
    id: 4,
    productName: 'Pants - blue',
    price: 599,
    size: 'medium',
    category: ProductCategory.CLOTHES,
    isRed: false,
    isBlue: true,
    isGreen: false,
  },
  {
    id: 5,
    productName: 'Pants - green',
    price: 599,
    size: 'medium',
    category: ProductCategory.CLOTHES,
    isRed: false,
    isBlue: false,
    isGreen: true,
  },
  {
    id: 6,
    productName: 'Pants - red',
    price: 599,
    size: 'medium',
    category: ProductCategory.CLOTHES,
    isRed: true,
    isBlue: false,
    isGreen: false,
  },
];
