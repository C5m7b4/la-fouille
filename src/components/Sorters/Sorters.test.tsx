import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import Sorters from './Sorters';
import { people } from '../../mockData/people';

describe('Sorters', () => {
  it('should render corretly', () => {
    const { asFragment } = render(
      <Sorters dataSource={people} setSortProperty={(sortProperty) => {}} />
    );

    expect(asFragment).toMatchSnapshot();
  });
  it('should use an empty object if no array is present', () => {
    const { asFragment } = render(
      <Sorters dataSource={[]} setSortProperty={(sortProperty) => {}} />
    );

    expect(() => screen.getByTestId('ss-firstName-true')).toThrowError(
      'Unable to find an element by: [data-testid="ss-firstName-true"]'
    );
  });
  it('should handle change event', () => {
    const { asFragment } = render(
      <Sorters dataSource={people} setSortProperty={(sortProperty) => {}} />
    );

    const selectElement: HTMLSelectElement =
      screen.getByTestId('sorter-select');
    fireEvent.change(selectElement, 2);
    expect(selectElement.value).toEqual('firstName-true');
  });
});
