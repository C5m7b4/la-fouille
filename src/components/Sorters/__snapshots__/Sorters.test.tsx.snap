// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorters should render corretly 1`] = `[Function]`;
